### PR TITLE
Fix CircleCI config to comply with Gradle docs

### PR DESCRIPTION
**Pull request by hmijail**
Signed-off-by: Horacio Mijail Anton Quiles <hmijail@gmail.com>

## PR description
Removing ineffective/redundant/not recommended settings:
* If parallel running is enabled in GRADLE_OPTS, it's not needed in the command lines. 
* Disabling the gradle daemon goes against the recommendations of the Gradle docs. 
* Setting -Xmx in GRADLE_OPTS is useless, according to Gradle docs.

I tried moving -Xmx somewhere more appropriate, but the purpose is not clear anyway. My understanding is that the goal is to limit memory usage to avoid Out Of Memory errors in CircleCI. But for that, the Gradle defaults (which are currently in effect, since the setting isn't being applied) are more strict, so it wouldn't be helpful to apply the setting without serious review.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -68,7 +68,7 @@ jobs:
       - run:
           name: Assemble
           command: |
-            ./gradlew --no-daemon --parallel clean spotlessCheck compileJava compileTestJava assemble
```
Comment (shemnon) [inline .circleci/config.yml:71]: Let's keep `--no-daemon` - daemons are useless in ci docker images.

**Comment by shemnon** [comment]
The gradle daemon recommendations are based on developer machines, not CI boxes.  Their daemon is of questionable value as some tools (such as JMH) don't work with the daemon runnning.

**Comment by hmijail** [comment]
AFAICT, Gradle recommends the daemon _specially_ in CI: https://blog.andresteingress.com/2018/09/05/gradle-daemon-and-jenkins.html 

My understanding is that the -Xmx switches in GRADLE_OPTS are always irrelevant, because the build and test VMs are always forked with different parameters that are set elsewhere, even without a daemon: https://docs.gradle.org/current/userguide/build_environment.html#sec:configuring_jvm_memory

File: .circleci/config.yml (hunk 1 above)
Comment (hmijail) [reply .circleci/config.yml:71]: Are these changes still requested, given that Gradle specially recommends the daemon in CI?
Should I add --no-daemon only to JMH?
At any rate I can confirm that acceptance tests work well without the daemon (and I'd say faster, but I haven't real measurements). And this build did finish correctly.

**Comment by shemnon** [comment]
First, I think you missed a subtlety in my response.  I'm fine dropping the -Xmx parts.

However, I'm not fine dropping the no-daemon.  The webpage you cite is not Gradle documentation, it's a blog by someone not employed by Gradle.  The blog was also talking about Jenkins for CI, not CircleCI.  It is also talking about a non-containerized build environment, whereas CircleCI is heavily containerized.  Gradle gets executed once per container before the container is torn down and brought up again, running the daemon actually would incur a performance penalty in that case starting up a second resident process, no JVM startup is avoided.  Also, check out the two comments to that blog.  The first questions how effective it is in a docker based CI solution (which is the same setup Circle CI uses).  The second pointing to the stability issues that the Gradle daemon introduces, which I myself have also encountered.

**Comment by hmijail** [comment]
Regarding `-Xmx`: 
Sorry, I was just trying to point out that the `-Xmx` thing is actually orthogonal to usage of the daemon. It took me time to understand the thing, the documentation is mediocre in Gradle and positively wrong in CircleCI (I opened issues with them already), and I thought it'd be helpful to pass the understanding around. But yes, that's agreed and "done".

Regarding the daemon: 
Please bear with me. My overarching point here is that it looks like we're accumulating cruft that no one knows why it's there, and it only makes it more difficult/scary to get in and clean up - particularly in the corners that people tend to avoid, like Gradle and CI. I'm trying to break what I think is a vicious circle.

I'm trying to remove cruft and simplify. In general Gradle knows what to do, so why not just let it do it?
In the case of the daemon, even with `--no-daemon` Gradle does start the daemon and kills it after the build finishes. 
```
$ export _JAVA_OPTS=-D_A_JVM_HAS_STARTED
$ ./gradlew clean build --no-daemon --no-build-cache 2>&1 | grep _A_JVM_HAS_STARTED | wc -l
     100
$ ./gradlew clean build --daemon --no-build-cache 2>&1 | grep _A_JVM_HAS_STARTED | wc -l
      97
```

In smaller tasks, like acceptanceTest, you'll see that in both cases the same number of JVMs are started. But I haven't seen a case where `--no-daemon` causes Gradle to start less JVMs.
(I have to admit that I expected the difference would be better towards the daemon. But even if there was no difference, my point remains)

I cited that page because it concisely links to the Gradle docs and to a (German) podcast with a **Gradle core dev**, where he additionally comments in the page and links to yet another [webcast](https://www.youtube.com/watch?v=OV-medORVsU&t=25m10s) with yet **another Gradle core dev**. 
Both core devs answer the daemon-on-CI question with yes. The second one particularly says that only in throwaway machines with single runs the daemon will not help you - but even then he doesn't say that it's better to disable it.

This comes close to splitting hairs, but that is my point: it is NOT worse, it MIGHT be better and it is cleaner, so why not just let Gradle do its thing?

Regarding the daemon instability: I've also had to deal with CI instability (all my PRs to this repo have been to help fix it), but in my case I managed to fix it by doing this kind of simplification and fixing underlying problems. Of course if you know that any part of the build only works with `--no-daemon`, then of course by all means let's use it locally and purposefully (and hopefully with a miniexplanation?), so it's clear that it has a reason to be there. But even then, note that in that webcast the Gradle dev says that they have gone out of their way with external plugin makers to make sure that they work correctly with the daemon because that's the way forward.

Having said all of that: at the end of the day I don't really mind one way or another because I don't generally work in Besu; this is just a backport of what was **needed** in my team's fork. So if you prefer it as it was, I'll leave it.

**Comment by shemnon** [comment]
These flags are not cruft, are not scary, are not areas people tend to avoid, and this is not a vicious cycle.  The flag was put there purposefully and with measured intent.

**Comment by hmijail** [comment]
If it was measured, I just gave a measure that negates it - and yet has just been brushed off without explanation.
If it wasn't a vicious cycle, the unmeasured add-ons wouldn't accumulate like cruft. 
If it wasn't scary, more people would be fixing this stuff instead of poking things around randomly and hoping that the problems go away.

I will re-add the --no-daemon.

**Pull request merged**